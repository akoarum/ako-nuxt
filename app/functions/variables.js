export const VALIDATES = {
  EMPTY: {
    message: '入力してください。'
  },
  NUMBER: {
    regex: /^[0-9]+?$/,
    message: '数字で入力してください。'
  },
  EMAIL: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスの形式になっていません。'
  },
  TEL: {
    regex: /^[0-9]{2,4}-?[0-9]{2,4}-?[0-9]{4}$/,
    message: '電話番号の形式になっていません。'
  },
  URL: {
    regex: /^http(s):\/\/.+?\.+?$/,
    message: 'URLの形式になっていません。'
  }
}
