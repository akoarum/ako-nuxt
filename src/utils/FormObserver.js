export default class FormValidator {
  constructor(names = []) {
    this._items = this._createItems(names)
  }

  /**
   * @return {Boolean}
   */
  get btnDisabled() {
    for (const name in this._items) {
      if (this._items[name].error || !this._items[name].dirty) return true
    }
    return false
  }

  /**
   * itemオブジェクトの生成
   * @param {Array} names name配列
   */
  _createItems(names) {
    const items = {}

    names.forEach(item => {
      items[item] = {
        error: false,
        dirty: false
      }
    })

    return items
  }

  /**
   * itemの更新
   * @param {String}  name  name値
   * @param {String}  prop  更新するプロパティ名
   * @param {Boolean} value 値
   */
  updateItem(name, prop, value) {
    this._items[name][prop] = value
  }
}
