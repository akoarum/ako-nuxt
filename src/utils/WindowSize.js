export default class WindowSize {
  constructor(delay = 80) {
    this._width = 0
    this._height = 0
    this._delay = delay
    this._cullTime = null
    this._initialized = false
    this._handler = this._handleResize.bind(this)
  }

  get width() {
    return this._width
  }

  get height() {
    return this._height
  }

  get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  initialize() {
    if (this._initialized) return this
    this._updateState()
    window.addEventListener('resize', this._handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  _updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  _handleResize() {
    clearTimeout(this._cullTime)
    this._cullTime = setTimeout(() => {
      this._updateState()
    }, this._delay)
  }

  /**
   * 解放する
   */
  destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this._handler)
    this._initialized = false
    return this
  }
}
