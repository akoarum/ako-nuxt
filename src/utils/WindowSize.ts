export default class WindowSize {
  private _width: number = 0
  private _height: number = 0
  private delay: number = 80
  private cullTime = null
  private _initialized: boolean = false
  private handler = this.handleResize.bind(this)

  constructor(delay: number = 80) {
    this.delay = delay
  }

  public get width() {
    return this._width
  }

  public get height() {
    return this._height
  }

  public get initialized() {
    return this._initialized
  }

  /**
   * 初期化メソッド
   */
  public initialize() {
    if (this._initialized) return this
    this.updateState()
    window.addEventListener('resize', this.handler)
    this._initialized = true
    return this
  }

  /**
   * ステートの更新
   */
  public updateState() {
    this._width = window.innerWidth
    this._height = window.innerHeight
  }

  /**
   * リサイズハンドラ
   */
  private handleResize() {
    clearTimeout(this.cullTime)
    this.cullTime = setTimeout(() => {
      this.updateState()
    }, this.delay)
  }

  /**
   * 解放する
   */
  public destroy() {
    if (!this._initialized) return this
    window.removeEventListener('resize', this.handler)
    this._initialized = false
    return this
  }
}
