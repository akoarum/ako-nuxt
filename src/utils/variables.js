export const VALIDATES = {
  required: '入力してください。',
  selected: '選択してください。',
  file: 'ファイルをアップロードしてください。',
  tel: {
    regex: /^\d{2,4}-?\d{4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
