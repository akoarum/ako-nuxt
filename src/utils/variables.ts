import { ValidatePattern } from '~/types/models'

export const VALIDATE_PATTERNS: { [key: string]: ValidatePattern } = {
  required: {
    message: '入力してください。'
  },
  selected: {
    message: '選択してください。'
  },
  tel: {
    regex: /^\d{2,5}-?\d{1,4}-?\d{4}$/,
    message: '電話番号は半角数字とハイフンで入力してください。'
  },
  email: {
    regex: /^[-a-zA-Z0-9!#$%&'*+\-/=?^_`.{|}~]+@[-a-zA-Z0-9]+(\.[-a-zA-Z0-9]+)*\.[a-zA-Z]{2,6}$/,
    message: 'メールアドレスを正しく入力してください。'
  }
}
