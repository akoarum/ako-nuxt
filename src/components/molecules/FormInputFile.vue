<template>
  <div class="formInputFile">
    <InputFile
      :name="name"
      :aria-invalid="String(!!error)"
      :aria-describedby="error ? `${name}-error` : false"
      class="formInputFile__input"
      @input="handleInput"
    />
    <VTexts
      v-if="error"
      :id="`${name}-error`"
      role="alert"
      aria-live="assertive"
      class="formInputFile__error"
      small
      caution
    >
      <span>{{ error }}</span>
    </VTexts>
  </div>
</template>

<script>
import { VALIDATES } from '~/utils/variables'
import { InputFile, VTexts } from '~/components/atoms'

export default {
  components: { InputFile, VTexts },
  props: {
    name: { type: String, required: true },
    value: { type: [File, null] },
    required: { type: Boolean }
  },
  data() {
    return {
      error: '',
      isDirty: false
    }
  },
  methods: {
    validate() {
      if (!this.required) return ''
      if (!this.isDirty) return ''
      if (this.value) return ''
      return VALIDATES.file
    },
    handleInput(value) {
      this.$emit('input', value)
      this.$nextTick(() => {
        this.isDirty = true
        this.error = this.validate()
      })
    }
  }
}
</script>

<style lang="scss" scoped>
.formInputFile__error {
  @include media-min {
    margin-top: 8px;
  }

  @include media-max {
    margin-top: 4px;
  }
}
</style>
