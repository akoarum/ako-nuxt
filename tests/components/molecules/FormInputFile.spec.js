import { shallowMount } from '@vue/test-utils'
import { VALIDATES } from '~/utils/variables'
import FormInputFile from '~/components/molecules/FormInputFile'

describe('FormInputFile', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(FormInputFile, {
      propsData: {
        name: 'test'
      }
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })
    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    describe('validate: バリデーション', () => {
      it('requiredでなければ空文字列を返す', () => {
        expect(wrapper.vm.validate()).toBe('')
      })

      it('isDirtyでなければ空文字列を返す', () => {
        wrapper.setProps({ required: true })
        expect(wrapper.vm.validate()).toBe('')
      })

      it('valueに値があれば空文字列を返す', () => {
        wrapper.setData({ isDirty: true })
        wrapper.setProps({ required: true, value: {} })
        expect(wrapper.vm.validate()).toBe('')
      })

      it('上記以外ならfileエラーを返す', () => {
        wrapper.setData({ isDirty: true })
        wrapper.setProps({ required: true, value: null })
        expect(wrapper.vm.validate()).toBe(VALIDATES.file)
      })
    })

    it('handleInput: 引数を持ってemit', () => {
      wrapper.vm.handleInput('test')
      expect(wrapper.emitted().input).toBeTruthy()
    })
  })
})
