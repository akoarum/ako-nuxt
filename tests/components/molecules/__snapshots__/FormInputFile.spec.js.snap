// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInputFile initialized snapshot 1`] = `
<div
  class="formInputFile"
>
  <inputfile-stub
    aria-invalid="false"
    class="formInputFile__input"
    label="アップロードする"
    name="test"
  />
   
  <!---->
</div>
`;
